### PR TITLE
L1T: Add jet quality and eg/tau isolation data vs. emu comparisons and ignore object index

### PR DESCRIPTION
**Pull request by bundocka**
#### PR description:

- Adds jet quality and eg/tau isolation to the CaloL2 data vs emulator plots in the online DQM
- Bug fix to ignore the ordering of CaloL2 objects for data vs emulator comparisons

#### PR validation:

- Ran locally on lxplus with recent ExpressPhysics data and checked histogram outputs are as expected
- Ran unit tests with `cmsRun DQM/Integration/python/clients/l1tstage2emulator_dqm_sourceclient-live_cfg.py unitTest=True`

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: DQM/L1TMonitorClient/src/L1TStage2CaloLayer2DEClientSummary.cc [hunk 1]
```diff
@@ -152,7 +155,7 @@ void L1TStage2CaloLayer2DEClientSummary::processHistograms(DQMStore::IGetter &ig
     // double totalJets = 0, goodJets = 0, jetPosOff = 0, jetEtOff = 0;
```
Comment (emanueleusai) [inline DQM/L1TMonitorClient/src/L1TStage2CaloLayer2DEClientSummary.cc:155]: Can remove these comments?

File: DQM/L1TMonitorClient/src/L1TStage2CaloLayer2DEClientSummary.cc [hunk 2]
```diff
@@ -152,7 +155,7 @@ void L1TStage2CaloLayer2DEClientSummary::processHistograms(DQMStore::IGetter &ig
...
 
     // by default show 0% agreement (for edge case when no objects are found)
-    double goodRatio = 0, posOffRatio = 0, etOffRatio = 0;
+    double goodRatio = 0, posOffRatio = 0, etOffRatio = 0, qualOffRatio = 0;
 
     // hist = jetSummary_->getTH1F();
     // newHist = jetSummary->getTH1F();
```
Comment (emanueleusai) [inline DQM/L1TMonitorClient/src/L1TStage2CaloLayer2DEClientSummary.cc:161]: can remove these comments?

File: DQM/L1TMonitorClient/src/L1TStage2CaloLayer2DEClientSummary.cc [hunk 3]
```diff
@@ -179,7 +185,7 @@ void L1TStage2CaloLayer2DEClientSummary::processHistograms(DQMStore::IGetter &ig
 
     // by default show 0% agreement (for edge case when no objects are found)
     double goodEgRatio = 0, egPosOffRatio = 0, egEtOffRatio = 0, goodIsoEgRatio = 0, isoEgPosOffRatio = 0,
-           isoEgEtOffRatio = 0;
+           isoEgEtOffRatio = 0, egIsoOffRatio = 0;
 
     // hist = egSummary_->getTH1F();
     // newHist = egSummary->getTH1F();
```
Comment (emanueleusai) [inline DQM/L1TMonitorClient/src/L1TStage2CaloLayer2DEClientSummary.cc:191]: Can remove these comments?

File: DQM/L1TMonitorClient/src/L1TStage2CaloLayer2DEClientSummary.cc [hunk 4]
```diff
@@ -221,7 +234,7 @@ void L1TStage2CaloLayer2DEClientSummary::processHistograms(DQMStore::IGetter &ig
 
     // by default show 0% agreement (for edge case when no objects are found)
     double goodTauRatio = 0, tauPosOffRatio = 0, tauEtOffRatio = 0, goodIsoTauRatio = 0, isoTauPosOffRatio = 0,
-           isoTauEtOffRatio = 0;
+           isoTauEtOffRatio = 0, tauIsoOffRatio = 0;
 
     // hist = tauSummary_->getTH1F();
     // newHist = tauSummary->getTH1F();
```
Comment (emanueleusai) [inline DQM/L1TMonitorClient/src/L1TStage2CaloLayer2DEClientSummary.cc:240]: Can remove these comments?

**Comment by emanueleusai** [comment]
@bundocka there are some comments (not introduced by you) that can be deleted. It's up to you if you want to implement these now or leave them to a future PR. Other than that the PR is good to go for me (DQM plot differences exactly where we expect them)

**Comment by emanueleusai** [comment]
please test

**Comment by emanueleusai** [comment]
I was just waiting for the tests to finish after the author removed the comments.

**Comment by emanueleusai** [comment]
+1

**Comment by perrotta** [comment]
With the small stat visible in the bot tests for the HI wf 140.56 the "problematic" histos are now empty, which is a effect of your fix to ignore the ordering. Do you see the same (i.e. no more unmatched jets) with the supposedly larger stat that you used for your tests? Please confirm that this is expected.

Why is that bad ordering present only for jets and egamma, and not for the other objects (tau, ...) dealt with by this code? I suppose there is some bug uphill for jets and egamma, which should get better fixed rather than patching the code downhill to cope with it.

After having verified that there is this ordering issue, I now wonder if it still makes sense to fill the "Emul" plots with the object pointed by what has been verified to be the wrong index, see for example [L481-486](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L481-L486) of L1TdeStage2CaloLayer2.cc. Since you have verified that Et, Pos, Qual must be the same, you should then fill the Emul quantities with the ones extracted from `dataIt`: am I correct? The same for the corresponding Egamma plots in [L660-687](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L660-L687).

Moreover, this is a minor point but if for any of the reasons above you have to add another commit to this PR, you could profit and simplify the code in order to make it better readable and more easily maintainable. For example the whole [L429-451](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L429-L451) could be more simply rewritten as
```C++
      // object pt mismatch
      bool etGood = dataIt->hwPt() == emulIt->hwPt();
      // object position mismatch
      bool posGood = dataIt->hwPhi() == emulIt->hwPhi() && dataIt->hwEta() == emulIt->hwEta();
      // object quality mismatch
      bool qualGood = dataIt->hwQual() == emulIt->hwQual();
```
And the same for L609-632, [L808-851](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L808-L851)

**Comment by perrotta** [comment]
hold
- @bundocka please take care of https://github.com/cms-sw/cmssw/pull/40511#issuecomment-1399212495

**Comment by bundocka** [comment]
> With the small stat visible in the bot tests for the HI wf 140.56 the "problematic" histos are now empty, which is a effect of your fix to ignore the ordering. Do you see the same (i.e. no more unmatched jets) with the supposedly larger stat that you used for your tests? Please confirm that this is expected.

Running over 36k events, there is perfect agreement for jets, however there are still a few unmatched EGs and ~30 unmatched taus. This is as expected. 

> Why is that bad ordering present only for jets and egamma, and not for the other objects (tau, ...) dealt with by this code? I suppose there is some bug uphill for jets and egamma, which should get better fixed rather than patching the code downhill to cope with it.

The bad ordering is present for jets, EG, and tau also. See the same fix to the tau matching here [L853-L866](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L853-L866). The mismatches arise when there are lots of low pt objects with the same et and eta, but with different phi, due to a very subtle difference in how the firmware and emulator order objects in phi. So occasionally, if there is more than 1 object with the same et and eta, but different phi, the firmware and emulator may pick different objects to be in the top 6 pt candidates for each eta half (we have a total of 12 candidates for each object, 6 for each of positive and negative eta). Quite some time was spent to try and implement the same logic in the emulator as the firmware, but it's very difficult, due to how the pipelined nature of the firmware constructs the objects as the calorimeter towers arrive into the FPGA logic. Rather than spend lots more time and effort, and make the emulator much more complicated, it was chosen to instead ignore the ordering of the objects in the list. 


> After having verified that there is this ordering issue, I now wonder if it still makes sense to fill the "Emul" plots with the object pointed by what has been verified to be the wrong index, see for example [L481-486](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L481-L486) of [[L1TdeStage2CaloLayer2.cc](http://l1tdestage2calolayer2.cc/)](http://l1tdestage2calolayer2.cc/). Since you have verified that Et, Pos, Qual must be the same, you should then fill the Emul quantities with the ones extracted from `dataIt`: am I correct? The same for the corresponding Egamma plots in [L660-687](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L660-L687).
> 

Thanks for spotting this. I have updated the code to take this into account. 

> Moreover, this is a minor point but if for any of the reasons above you have to add another commit to this PR, you could profit and simplify the code in order to make it better readable and more easily maintainable. For example the whole [L429-451](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L429-L451) could be more simply rewritten as
> 
> ```c++
>       // object pt mismatch
>       bool etGood = dataIt->hwPt() == emulIt->hwPt();
>       // object position mismatch
>       bool posGood = dataIt->hwPhi() == emulIt->hwPhi() && dataIt->hwEta() == emulIt->hwEta();
>       // object quality mismatch
>       bool qualGood = dataIt->hwQual() == emulIt->hwQual();
> ```
> 
> And the same for L609-632, [L808-851](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L808-L851)

Thanks, I've implemented your suggestion.

**Comment by perrotta** [comment]
please test

**Comment by emanueleusai** [comment]
please test

**Comment by emanueleusai** [comment]
type l1t

**Comment by perrotta** [comment]
unhold

**Comment by emanueleusai** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**